src/particles/webgl/ParticleRenderer.js

```diff
@@ -387,16 +389,21 @@ export default class ParticleRenderer extends core.ObjectRenderer
      * @param {number} stride - Stride to use for iteration.
      * @param {number} offset - Offset to start at.
      */
-    uploadAlpha(children, startIndex, amount, array, stride, offset)
+    uploadTint(children, startIndex, amount, array, stride, offset)
     {
-        for (let i = 0; i < amount; i++)
+        for (let i = 0; i < amount; ++i)
         {
-            const spriteAlpha = children[startIndex + i].alpha;
-
-            array[offset] = spriteAlpha;
-            array[offset + stride] = spriteAlpha;
-            array[offset + (stride * 2)] = spriteAlpha;
-            array[offset + (stride * 3)] = spriteAlpha;
+            const sprite = children[startIndex + i];
+            const premultiplied = sprite._texture.baseTexture.premultipliedAlpha;
+            const alpha = sprite.alpha;
+            // we dont call extra function if alpha is 1.0, that's faster
+            const argb = alpha < 1.0 && premultiplied ? premultiplyTint(sprite._tintRGB, alpha)
+                : sprite._tintRGB + (alpha * 255 << 24);
+
+            array[offset] = argb;
+            array[offset + stride] = argb;
+            array[offset + (stride * 2)] = argb;
+            array[offset + (stride * 3)] = argb;
```

Comment:
Looking this and the other functions, I'm not clear why there are 4 assignments in them? If the stride is 1, and you are uploading UNSIGNED_BYTE to a vec4, wouldn't you only assign `array[offset]`?

Why do you have to fill 4 Uint32s for an UNSIGNED_BYTE vec4? This pattern is used in all the upload funcs, just not sure why each of them copy the data 4 times?

Reply:
That's how our SpriteRenderer works. 4 assignments are for different vertices, and its not clear what stride is, its always more than zero. This abstraction exists because we dont know which fields are static and which are dynamic.

Reply:
Ah, that's right we upload it per vert (one for each uv coord). Forgot about that! Thanks!